src/modules/flow/flower-power/flower-power.c

```diff
@@ -0,0 +1,709 @@
+    struct sol_drange *fertilizer, struct sol_drange *light,
+    struct sol_drange *temperature, struct sol_drange *water)
+
+{
+    struct packet_type_flower_power_packet_data packet_type_flower_power;
+    int ret;
+
+    SOL_NULL_CHECK(packet, -EINVAL);
+    if (sol_flow_packet_get_type(packet) != PACKET_TYPE_FLOWER_POWER)
+        return -EINVAL;
+
+    ret = sol_flow_packet_get(packet, &packet_type_flower_power);
+    SOL_INT_CHECK(ret, != 0, ret);
+
+    if (id)
+        *id = packet_type_flower_power.id;
+    if (timestamp)
+        *timestamp = packet_type_flower_power.timestamp;
+    if (fertilizer)
+        *fertilizer = packet_type_flower_power.fertilizer;
+    if (light)
+        *light = packet_type_flower_power.light;
+    if (temperature)
+        *temperature = packet_type_flower_power.temperature;
+    if (water)
+        *water = packet_type_flower_power.water;
+
+    return ret;
+}
+
+static int
+send_packet_type_flower_power_packet(struct sol_flow_node *src,
+    uint16_t src_port, char *id, char *timestamp,
+    struct sol_drange *fertilizer, struct sol_drange *light,
+    struct sol_drange *temperature, struct sol_drange *water)
+{
+    struct sol_flow_packet *packet;
+    int ret;
+
+    packet = packet_new_packet_type_flower_power(id, timestamp,
+        fertilizer, light, temperature, water);
+    SOL_NULL_CHECK(packet, -ENOMEM);
+
+    ret = sol_flow_send_packet(src, src_port, packet);
+    if (ret != 0)
+        sol_flow_packet_del(packet);
+
+    return ret;
+}
+
+struct http_get_data {
+    struct sol_flow_node *node;
+    char *client_id;
+    char *client_secret;
+    char *username;
+    char *password;
+    char *token;
+};
+
+struct filter_data {
+    char *id;
+};
+
+static int
+http_get_open(struct sol_flow_node *node, void *data, const struct sol_flow_node_options *options)
+{
+    struct http_get_data *mdata = data;
+    const struct sol_flow_node_type_flower_power_http_get_options *opts;
+
+    SOL_FLOW_NODE_OPTIONS_SUB_API_CHECK(options,
+        SOL_FLOW_NODE_TYPE_FLOWER_POWER_HTTP_GET_OPTIONS_API_VERSION,
+        -EINVAL);
+    opts = (const struct sol_flow_node_type_flower_power_http_get_options *)options;
+
+    if (!opts->client_id || !strlen(opts->client_id) ||
+        !opts->client_secret || !strlen(opts->client_secret)) {
+        SOL_ERR("Valid client id and secret are required.");
+        return -EINVAL;
+    }
+
+    mdata->client_id = strdup(opts->client_id);
+    SOL_NULL_CHECK(opts->client_id, -ENOMEM);
+    mdata->client_secret = strdup(opts->client_secret);
+    SOL_NULL_CHECK(opts->client_secret, -ENOMEM);
+
+    mdata->node = node;
+
+    return 0;
+}
+
+static void
+http_get_close(struct sol_flow_node *node, void *data)
+{
+    struct http_get_data *mdata = data;
+
+    free(mdata->client_id);
+    /* Zero memory used to store client secret and password */
+    if (mdata->client_secret) {
+        memset(mdata->client_secret, 0, strlen(mdata->client_secret) + 1);
```

Comment:
client_secret must be in memory while this node is in use, because it's required to create a new token all the time a new username is used (or token expires (not handled)).

I'm just zeroing it after we're done with it.